src/targets/gpu/include/migraphx/gpu/context.hpp

```diff
@@ -272,8 +274,40 @@ struct context
 
         auto v_streams        = v.at("streams");
         std::size_t n_streams = v_streams.without_key().to<std::size_t>();
+        this->current_device  = std::make_shared<hip_device>(0, n_streams);
+#ifdef MIGRAPHX_HAS_FIND_2_API
+        {
+            auto v_gfx_arch = v.at("gfx_arch").to<std::string>();
+
+            auto v_cu_count        = v.at("cu_count");
+            std::size_t n_cu_count = v_cu_count.without_key().to<std::size_t>();
 
-        this->current_device = std::make_shared<hip_device>(0, n_streams);
+            std::string v_miopen_version = v.at("miopen_version").to<std::string>();
+
+            std::string current_gfx_arch       = this->current_device->get_device_name();
+            std::size_t current_cu_count       = this->current_device->get_cu_count();
+            std::string current_miopen_version = get_miopen_version();
+            if(n_cu_count != current_cu_count or v_gfx_arch != current_gfx_arch)
+            {
+                std::cerr << "MIGraphX model was compiled for gfx_arch: " << v_gfx_arch
+                          << " with number of CUs=" << n_cu_count
+                          << ", but current device has gfx_arch: " << current_gfx_arch
+                          << " with number of CUs=" << current_cu_count
+                          << ", performance may suffer. \
+                Consider re-compiling the model with environment variable MIOPEN_FIND_ENFORCE=3 to re-tune the model."
+                          << std::endl;
+            }
+            if(current_miopen_version != v_miopen_version)
```

Comment:
If we are saving the tuning information then we shouldn't be warning the user about a different miopen version, because we have already saved the tuning information(with find 2.0) even if its not in miopen's db.

Reply:
If the user is using a later version of MIOpen then, it is possible that that would get a performance boost if they were to recompile. 

If user is using older version then also it is possible that with the saved tuning informations, they wouldn't get the best performance.

Reply:
I dont think the perf difference is big enough that we should need to warn the user.

It could be also they tuned with an older version because they know its faster so the messages would even be annoying.

Reply:
This could also be an application so the user doesnt even know how to recompile the model.

Reply:
Saved tuning information would be specific to kernel implementation in a specific MIOpen version. If the kernel's implementation is different across MIOpen versions (i am not sure how likely it is). such that they take a different number of tuning parameters or let's say not use all of previously saved tuning parameters then, in that case, it is not clear what sort of results user would see. 

If user finds that their model works better without MIOPEN_FIND_ENFORCE=3 then that choice is better left to them. We should warn them about possible reasons why they wouldn't see best performance.

Reply:
> such that they take a different number of tuning parameters or let's say not use all of previously saved tuning parameters then, in that case, it is not clear what sort of results user would see.

We should have an error from MIOpen for that(ie miopenStatusVersionMismatch). If not, then we need to open an issue with MIOpen. We should also throw an error when MIOpen returns `miopenStatusVersionMismatch`, but this should be a hard error not a warning.

> We should warn them about possible reasons why they wouldn't see best performance.

Only with an env var to enable it. We should not spam the terminal with this during application time, with no way to turn it off. This could interfere with an application(especially if this is an interactive bash process such as bash completion or fuzzy searching history, etc).